code/modules/asset_cache/spritesheet/legacy/legacy_spritesheet.dm

```diff
@@ -0,0 +1,418 @@
+	for(var/size_id in sizes)
+		var/size = sizes[size_id]
+		var/file_path = size[SPRSZ_STRIPPED]
+		var/file_hash = rustg_hash_file(RUSTG_HASH_MD5, file_path)
+		SSassets.transport.register_asset("[name]_[size_id].png", file_path, file_hash=file_hash)
+	var/css_name = "spritesheet_[name].css"
+	var/file_directory = "data/spritesheets/[css_name]"
+	fdel(file_directory)
+	var/css = generate_css()
+	rustg_file_write(css, file_directory)
+	var/css_hash = rustg_hash_string(RUSTG_HASH_MD5, css)
+	SSassets.transport.register_asset(css_name, fcopy_rsc(file_directory), file_hash=css_hash)
+
+	if(CONFIG_GET(flag/save_spritesheets))
+		save_to_logs(file_name = css_name, file_location = file_directory)
+
+	fdel(file_directory)
+
+	if (CONFIG_GET(flag/cache_assets) && cross_round_cachable)
+		write_to_cache()
+	fully_generated = TRUE
+	// If we were ever in there, remove ourselves
+	SSasset_loading.dequeue_asset(src)
+
+/datum/asset/spritesheet/queued_generation()
+	realize_spritesheets(yield = TRUE)
+
+/datum/asset/spritesheet/ensure_ready()
+	if(!fully_generated)
+		realize_spritesheets(yield = FALSE)
+	return ..()
+
+/datum/asset/spritesheet/send(client/client)
+	if (!name)
+		return
+
+	if (!should_refresh())
+		return send_from_cache(client)
+
+	var/all = list("spritesheet_[name].css")
+	for(var/size_id in sizes)
+		all += "[name]_[size_id].png"
+	. = SSassets.transport.send_assets(client, all)
+
+/datum/asset/spritesheet/get_url_mappings()
+	if (!name)
+		return
+
+	if (!should_refresh())
+		return get_cached_url_mappings()
+
+	. = list("spritesheet_[name].css" = SSassets.transport.get_asset_url("spritesheet_[name].css"))
+	for(var/size_id in sizes)
+		.["[name]_[size_id].png"] = SSassets.transport.get_asset_url("[name]_[size_id].png")
+
+/datum/asset/spritesheet/proc/ensure_stripped(sizes_to_strip = sizes)
+	for(var/size_id in sizes_to_strip)
+		var/size = sizes[size_id]
+		if (size[SPRSZ_STRIPPED])
+			continue
+
+		// save flattened version
+		var/png_name = "[name]_[size_id].png"
+		var/file_directory = "data/spritesheets/[png_name]"
+		fcopy(size[SPRSZ_ICON], file_directory)
+		var/error = rustg_dmi_strip_metadata(file_directory)
+		if(length(error))
+			stack_trace("Failed to strip [png_name]: [error]")
+		size[SPRSZ_STRIPPED] = icon(file_directory)
+
+		// this is useful here for determining if weird sprite issues (like having a white background) are a cause of what we're doing DM-side or not since we can see the full flattened thing at-a-glance.
+		if(CONFIG_GET(flag/save_spritesheets))
+			save_to_logs(file_name = png_name, file_location = file_directory)
+
+		fdel(file_directory)
+
+/datum/asset/spritesheet/proc/generate_css()
+	var/list/out = list()
+
+	for (var/size_id in sizes)
+		var/size = sizes[size_id]
+		var/icon/tiny = size[SPRSZ_ICON]
+		out += ".[name][size_id]{display:inline-block;width:[tiny.Width()]px;height:[tiny.Height()]px;background-image:url('[get_background_url("[name]_[size_id].png")]');background-repeat:no-repeat;}"
```

Comment:
I'm thinking maybe it would help to use `get_icon_dimensions` to reduce overhead, unless it's been either disproved or the icons are being constantly generated and not cached (which I heavily doubt).